src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -174,84 +204,107 @@ private BoundSequence VisitIndexOrRangePatternIndexerAccess(BoundIndexOrRangePat
+                var indexerAccess = (BoundIndexerAccess)node.IndexerAccess;
+                return (BoundSequence)F.Sequence(
+                    ImmutableArray.Create<LocalSymbol>(receiverLocal.LocalSymbol),
+                    ImmutableArray.Create<BoundExpression>(receiverStore),
+                    MakeIndexerAccess(
+                        node.Syntax,
+                        receiverLocal,
+                        indexerAccess.Indexer,
+                        ImmutableArray.Create<BoundExpression>(integerArgument),
+                        default,
+                        default,
+                        expanded: false,
+                        argsToParamsOpt: default,
+                        defaultArguments: default,
+                        indexerAccess.Indexer.Type,
+                        oldNodeOpt: null,
+                        isLeftOfAssignment));
+            }
+            else
+            {
+                Debug.Assert(node.ArgumentPlaceholders.Length == 1);
+                var argumentPlaceholder = node.ArgumentPlaceholders[0];
+
+                AddPlaceholderReplacement(argumentPlaceholder, integerArgument);
+                var rewrittenIndexerAccess = VisitExpression(node.IndexerAccess);
+                RemovePlaceholderReplacement(argumentPlaceholder);
+                RemovePlaceholderReplacement(receiverPlaceholder);
+
+                return (BoundSequence)F.Sequence(
+                    ImmutableArray.Create<LocalSymbol>(receiverLocal.LocalSymbol),
+                    ImmutableArray.Create<BoundExpression>(receiverStore),
+                    rewrittenIndexerAccess);
+            }
         }
 
         /// <summary>
-        /// Used to construct a pattern index offset expression, of the form
+        /// Used to construct a pattern index offset expression (of type Int32), of the form
         ///     `unloweredExpr.GetOffset(lengthAccess)`
         /// where unloweredExpr is an expression of type System.Index and the
         /// lengthAccess retrieves the length of the indexing target.
         /// </summary>
         /// <param name="unloweredExpr">The unlowered argument to the indexing expression</param>
-        /// <param name="lengthAccess">
+        /// <param name="loweredLengthAccess">
         /// An expression accessing the length of the indexing target. This should
         /// be a non-side-effecting operation.
         /// </param>
         /// <param name="usedLength">
         /// True if we were able to optimize the <paramref name="unloweredExpr"/>
-        /// to use the <paramref name="lengthAccess"/> operation directly on the receiver, instead of
+        /// to use the <paramref name="loweredLengthAccess"/> operation directly on the receiver, instead of
         /// using System.Index helpers.
         /// </param>
         private BoundExpression MakePatternIndexOffsetExpression(
```

Comment:
>MakePatternIndexOffsetExpression [](http://example.com/codeflow?start=32&length=32)

This code got changed significantly in `main`. I do not think it makes sense to make changes here and then deal with merge conflicts. I think implementation strategy is likely to be significantly affected. I suggest we merge `main` first, then make necessary changes to the lowering based on the refactorings made in this PR.